### PR TITLE
feat: support validation options specifically for disabling pattern validation

### PR DESCRIPTION
**Pull request by TristanSpeakEasy**
addresses: https://github.com/getkin/kin-openapi/issues/409

### CODE REVIEW

File: openapi3/schema_validation_settings.go [hunk 1]
```diff
@@ -28,10 +30,19 @@ func MultiErrors() SchemaValidationOption {
 func VisitAsRequest() SchemaValidationOption {
 	return func(s *schemaValidationSettings) { s.asreq, s.asrep = true, false }
 }
+
 func VisitAsResponse() SchemaValidationOption {
 	return func(s *schemaValidationSettings) { s.asreq, s.asrep = false, true }
 }
 
+func DisableFormatValidation() SchemaValidationOption {
```
Comment (fenollp) [inline openapi3/schema_validation_settings.go:38]: You should flip the default (and mention this in README) so maybe name it `EnableFormatValidation`.

Please add some documentation, such as "EnableFormatValidation setting makes Validate not return an error when validating documents that mention schema formats that are not defined by the OpenAPIv3 specification".

**Review by fenollp — CHANGES_REQUESTED** [review]

File: openapi3/validation_issue409_test.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+	"github.com/getkin/kin-openapi/openapi3"
+)
+
+func TestIssue409(t *testing.T) {
+	l := openapi3.NewLoader()
+	s, err := l.LoadFromFile("testdata/issue409.yml")
+	require.NoError(t, err)
+
+	err = s.Validate(l.Context, openapi3.DisableSchemaPatternValidation())
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/validation_issue409_test.go:17]: Additional test added testing error is returned if option not provided

File: openapi3/testdata/issue409.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+          description: A developer-facing error message.
+          type: string
+        status_code:
+          description: The HTTP status code
+          format: int32
+          type: integer
+      type: object
+      required:
+        - message
+        - status_code
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/testdata/issue409.yml]: Apologies I had just grabbed a spec I was using for testing elsewhere, have now truncated this by quite a bit

File: openapi3/validation_options.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+type validationOptions struct {
+	schemaFormatValidationDisabled  bool
+	schemaPatternValidationDisabled bool
+}
+
+type contextKey string
+
+const (
+	contextKeyValidationOptions contextKey = "validationOptions"
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/validation_options.go]: 👍 made the suggested change

File: openapi3/schema.go [hunk 5]
```diff
@@ -36,9 +36,6 @@ var (
 	// SchemaErrorDetailsDisabled disables printing of details about schema errors.
 	SchemaErrorDetailsDisabled = false
 
-	//SchemaFormatValidationDisabled disables validation of schema type formats.
-	SchemaFormatValidationDisabled = false
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/schema.go:40]: Have added a breaking change line to the Readme, assuming this PR will be in the next release `v0.101.0` but please let me know if you would like me to change the version number

File: openapi3/loader.go [hunk 6]
```diff
@@ -51,7 +51,9 @@ type Loader struct {
 
 // NewLoader returns an empty Loader
 func NewLoader() *Loader {
-	return &Loader{}
+	return &Loader{
+		Context: context.Background(),
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/loader.go]: Change made

File: openapi3/validation_options.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package openapi3
+
+import "context"
+
+type ValidationOption func(options *validationOptions)
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/validation_options.go:6]: Added

File: openapi3/validation_options.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+	schemaPatternValidationDisabled bool
+}
+
+type contextKey string
+
+const (
+	contextKeyValidationOptions contextKey = "validationOptions"
+)
+
+func DisableSchemaFormatValidation() ValidationOption {
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/validation_options.go]: Added and semantics flipped

File: openapi3/validation_options.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+	return func(options *validationOptions) {
+		options.schemaFormatValidationDisabled = true
+	}
+}
+
+func DisableSchemaPatternValidation() ValidationOption {
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/validation_options.go:23]: Added

File: openapi3/schema_validation_settings.go [hunk 10]
```diff
@@ -28,10 +30,19 @@ func MultiErrors() SchemaValidationOption {
...
+	return func(s *schemaValidationSettings) { s.formatValidationDisabled = true }
+}
+
+func DisablePatternValidation() SchemaValidationOption {
```
[the remark this answers is not in the document]
Comment (TristanSpeakEasy) [reply openapi3/schema_validation_settings.go:44]: Added

File: openapi3/schema_validation_settings.go (hunk 1 above)
Comment (TristanSpeakEasy) [reply openapi3/schema_validation_settings.go]: Flipped and documentation added

**Comment by TristanSpeakEasy** [comment]
@fenollp thank you for the feedback, very much appreciated. Have addressed all the current comments

File: openapi3/schema_test.go [hunk 11]
```diff
@@ -1028,7 +1028,10 @@ func testType(t *testing.T, example schemaTypeExample) func(*testing.T) {
 		}
 		for _, typ := range example.AllInvalid {
 			schema := baseSchema.WithFormat(typ)
-			err := schema.Validate(context.Background())
+			ctx := withValidationOptions(context.Background(), &validationOptions{
```
Comment (fenollp) [inline openapi3/schema_test.go:1031]: Hm this needs a public API, surely users will want to be able to pass validation options to the `Validate` methods implemented to other types than `openapi3.T`.

**Review by fenollp — CHANGES_REQUESTED** [review]
Alright one last point and LGTM! Great work :)

File: openapi3/validation_issue409_test.go [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+func TestIssue409PatternNotIgnored(t *testing.T) {
+	l := openapi3.NewLoader()
+	s, err := l.LoadFromFile("testdata/issue409.yml")
+	require.NoError(t, err)
+
+	err = s.Validate(l.Context)
+	assert.Error(t, err)
+}
```
Comment (fenollp) [inline openapi3/validation_issue409_test.go:28]: Ah I'd also like to see 2 more tests that ensure hygienic use of `ctx`:
```go
func TestIssue409HygienicUseOfCtx(t *testing.T) {
	l := openapi3.NewLoader()
	doc, err := l.LoadFromFile("testdata/issue409.yml")
	require.NoError(t, err)

	err = doc.Validate(l.Context, openapi3.DisableSchemaPatternValidation())
	assert.NoError(t, err)
	err = doc.Validate(l.Context)
	assert.Error(t, err)

	// and the other way

	l = openapi3.NewLoader()
	doc, err = l.LoadFromFile("testdata/issue409.yml")
	require.NoError(t, err)

	err = doc.Validate(l.Context)
	assert.Error(t, err)
	err = doc.Validate(l.Context, openapi3.DisableSchemaPatternValidation())
	assert.NoError(t, err)
}
```

File: openapi3/schema_test.go (hunk 11 above)
Comment (TristanSpeakEasy) [reply openapi3/schema_test.go]: addressed

**Comment by TristanSpeakEasy** [comment]
@fenollp thanks for the feedback. Have addressed your comments. Also recently sent you an email (just in case it got caught up in spam)

**Review by fenollp — APPROVED** [review]

**Comment by fenollp** [comment]
Great work! Thanks a lot!

**Pull request merged**